File: P3_Longest_Substring_Without_Repeating_Characters/Python/my_imp/solution.py

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Two Pointers + Sliding Window | Time: O(n) | Space: O(n)
        left_ptr  = 0
        hash_map  = {}
        longest_window_size = 0

        for right_ptr, each_char in enumerate(s):
            if each_char in hash_map:
                left_ptr = max(left_ptr, hash_map[each_char] + 1)

            hash_map[each_char] = right_ptr
            longest_window_size = max(longest_window_size, right_ptr-left_ptr+1)

        return longest_window_size

File: P3_Longest_Substring_Without_Repeating_Characters/CPlusPlus/my_imp/solution.cpp
#include <map>
#include <vector>
#include <cstdio>
#include <solution.h>
#include <cstring>
#include <string>

#define MAX_CHARS_NUM 128

int Solution::lengthOfLongestSubstring(std::string s){
    // Two Pointers + Sliding Window | Time: O(n) | Space: O(n)
    int left_ptr  = 0;
    std::vector<int> hash_map(128, -1);//keep the latest position of occured char.
    int longest_window_size = 0;

    for (size_t right_ptr = 0; right_ptr < s.length(); ++right_ptr){
        left_ptr = std::max(left_ptr, hash_map[(int)s[right_ptr]]+1);
        hash_map[(int)s[right_ptr]] = (int)right_ptr;
        longest_window_size = std::max(longest_window_size, (int)right_ptr-left_ptr+1);
    }

    return longest_window_size;
}
